### PR TITLE
Fix and unit test for #804: Race condition in LanguageServerWrapper

### LINKED ISSUE #804: Race condition in LanguageServerWrapper on exception in StreamConnectionProvider.start()
Hi everyone,

I have observed the following behavior when I use a custom StreamConnectionProvider that tries to contact a remote language server using websockets: When the server is down, eventually lsp4e stops trying to recreate the language server connection. This is caused by an exception thrown in StreamConnectionProvider.start() that causes the following race condition:

- In https://github.com/eclipse/lsp4e/blob/9cdd9be00121f8c7fc386763e886d4e6a5f98787/org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java#L326 the start()-exception is handled by calling stop(), which sets `inititalizeFuture = null`
- In https://github.com/eclipse/lsp4e/blob/9cdd9be00121f8c7fc386763e886d4e6a5f98787/org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java#L259 initializeFuture is assigned the value from `CompletableFuture.supplyAsync`, which might happen before or after the start()-Exception has been handled.

Sometimes, the initializeFuture gets assigned the failed future from line 259 after the stop-Method has been called in line 326. When this happens, LanguageServerWrapper.start() does not start a new language server any more, because it only checks for initializeFuture == null in https://github.com/eclipse/lsp4e/blob/9cdd9be00121f8c7fc386763e886d4e6a5f98787/org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java#L256.

One possible solution might be to amend line 256 with something like `&& !this.initializeFuture.isCompletedExceptionally()`.

### PR DESCRIPTION
**Pull request by ChristophKaser**
Fixes #804

### CODE REVIEW

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java [hunk 1]
```diff
@@ -322,10 +322,13 @@ public synchronized void start() throws IOException {
 				FileBuffers.getTextFileBufferManager().addFileBufferListener(fileBufferListener);
 			}).exceptionally(e -> {
 				LanguageServerPlugin.logError(e);
-				initializeFuture.completeExceptionally(e);
 				stop();
-				return null;
+				throw new RuntimeException(e);
```
Comment (rubenporras) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java:326]: is there a reason for now just `throw e`?

Comment (ChristophKaser) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java:326]: Yes, unfortunately e is of type `Throwable` and `Function.apply` can only throw an unchecked exception.

**Review by mickaelistria — APPROVED** [review]
This looks good to me. I leave the final say (and merge if possible) to @rubenporras

**Review by rubenporras — APPROVED** [review]

**Comment by rubenporras** [comment]
thanks

**Comment by ChristophKaser** [comment]
Thanks for merging!

**Pull request merged**